qiskit/aqua/operators/state_functions/circuit_state_fn.py

```diff
@@ -0,0 +1,348 @@
+                                  is_measurement=self.is_measurement,
+                                  coeff=self.coeff * other.coeff)
+
+        if isinstance(other, CircuitStateFn) and self.is_measurement:
+            from .. import Zero
+            return self.compose(CircuitOp(other.primitive,
+                                          other.coeff)).compose(Zero ^ self.num_qubits)
+
+        from qiskit.aqua.operators import ComposedOp
+        return ComposedOp([new_self, other])
+
+    def tensor(self, other: OperatorBase) -> OperatorBase:
+        """ Tensor product
+        Note: You must be conscious of Qiskit's big-endian bit printing convention.
+        Meaning, Plus.tensor(Zero)
+        produces a |+⟩ on qubit 0 and a |0⟩ on qubit 1, or |+⟩⨂|0⟩, but would produce
+        a QuantumCircuit like
+        |0⟩--
+        |+⟩--
+        Because Terra prints circuits and results with qubit 0 at the end of the string or circuit.
+        """
+        # TODO accept primitives directly in addition to PrimitiveOp?
+
+        if isinstance(other, CircuitStateFn):
+            new_qc = QuantumCircuit(self.num_qubits + other.num_qubits)
+            # NOTE!!! REVERSING QISKIT ENDIANNESS HERE
+            new_qc.append(other.primitive, new_qc.qubits[0:other.primitive.num_qubits])
+            new_qc.append(self.primitive, new_qc.qubits[other.primitive.num_qubits:])
+            # TODO Fix because converting to dag just to append is nuts
+            # TODO Figure out what to do with cbits?
+            return CircuitStateFn(new_qc.decompose().to_instruction(),
+                                  coeff=self.coeff * other.coeff)
+        # pylint: disable=cyclic-import,import-outside-toplevel
+        from qiskit.aqua.operators import TensoredOp
+        return TensoredOp([self, other])
+
+    def to_density_matrix(self, massive: bool = False) -> np.ndarray:
+        """ Return numpy matrix of density operator, warn if more than 16 qubits to
+        force the user to set
+        massive=True if they want such a large matrix. Generally big methods like this
+        should require the use of a
+        converter, but in this case a convenience method for quick hacking and access
+        to classical tools is
+        appropriate. """
+
+        if self.num_qubits > 16 and not massive:
+            # TODO figure out sparse matrices?
+            raise ValueError(
+                'to_matrix will return an exponentially large matrix,'
+                ' in this case {0}x{0} elements.'
+                ' Set massive=True if you want to proceed.'.format(2 ** self.num_qubits))
+
+        # TODO handle list case
+        # Rely on VectorStateFn's logic here.
+        return StateFn(self.primitive.to_matrix() * self.coeff).to_density_matrix()
+
+    def to_matrix(self, massive: bool = False) -> np.ndarray:
+        """
+        NOTE: THIS DOES NOT RETURN A DENSITY MATRIX, IT RETURNS A CLASSICAL MATRIX CONTAINING
```

Comment:
Since there is the `to_density_matrix` also in this class does the Note need to be so "bold" as such. It can simply state what it does and even refer to :meth:`to_density_matrix` if that is what one was looking for.